xtc/src/ledger.rs

```diff
@@ -140,63 +193,127 @@ impl Ledger {
-    msg: "Insufficient Allowance",
-    code: APIError::InsufficientAllowance,
-};
+    let ledger = ic.get_mut::<Ledger>();
+    let amount_u64: u64 =
+        utils::convert_nat_to_u64(amount).expect("Amount cannot be represented as u64");
+    let fee = compute_fee(amount_u64);
 
-static UnknownError: ErrorDetails = ErrorDetails {
-    msg: "Unknown",
-    code: APIError::Unknown,
-};
+    ledger.approve(&caller, &to, amount_u64, fee)?;
 
-#[query]
-fn totalSupply() -> Nat {
-    StatsData::get().supply
+    let transaction = Transaction {
+        timestamp: ic.time(),
+        cycles: amount_u64,
+        fee,
+        kind: TransactionKind::Approve {
+            from: caller,
+            to: to,
+        },
+    };
+
+    Ok(Nat::from(
+        ic.get_mut::<HistoryBuffer>().push(transaction.clone()),
+    ))
 }
 
-#[update]
-pub async fn balance(account: Option<Principal>) -> u64 {
+#[update(name=transferErc20)]
+pub async fn transfer_erc20(to: Principal, amount: Nat) -> TxReceipt {
+    IsShutDown::guard();
+
     let ic = get_context();
     let caller = ic.caller();
+
     crate::progress().await;
-    let ledger = ic.get::<Ledger>();
-    ledger.balance(&account.unwrap_or(caller))
-}
 
-#[derive(Deserialize, CandidType)]
-pub struct TransferArguments {
-    pub to: Principal,
-    pub amount: u64,
-    // TODO(qt3ie) Notify argument.
+    let ledger = ic.get_mut::<Ledger>();
+    let amount_u64: u64 =
+        utils::convert_nat_to_u64(amount).expect("transfer failed - unable to convert amount");
+    let fee = compute_fee(amount_u64);
+    ledger.transfer(&caller, &to, amount_u64, fee)?;
+
+    let transaction = Transaction {
+        timestamp: ic.time(),
+        cycles: amount_u64,
+        fee,
+        kind: TransactionKind::Transfer {
+            from: caller,
+            to: to,
+        },
+    };
+
+    Ok(Nat::from(
+        ic.get_mut::<HistoryBuffer>().push(transaction.clone()),
+    ))
 }
 
-#[derive(CandidType, Debug)]
-pub enum TransferError {
-    InsufficientBalance,
-    AmountTooLarge,
-    CallFailed,
-    Unknown,
+#[update(name=transferFrom)]
+pub async fn transfer_from(from: Principal, to: Principal, amount: Nat) -> TxReceipt {
+    IsShutDown::guard();
+
+    let ic = get_context();
+    let caller = ic.caller();
+
+    crate::progress().await;
+
+    let ledger = ic.get_mut::<Ledger>();
+    let amount_u64: u64 =
+        utils::convert_nat_to_u64(amount).expect("transfer failed - unable to convert amount");
+    let fee = compute_fee(amount_u64);
+    ledger.transfer_from(&caller, &from, &to, amount_u64, fee)?;
+
+    let transaction = Transaction {
+        timestamp: ic.time(),
+        cycles: amount_u64,
+        fee,
+        kind: TransactionKind::TransferFrom {
+            caller: caller,
+            from: from,
+            to: to,
+        },
+    };
+
+    Ok(Nat::from(
+        ic.get_mut::<HistoryBuffer>().push(transaction.clone()),
+    ))
```

Comment:
warning: use of deprecated function `ic_kit::get_context`: get_context is deprecated use ic_kit::ic::*

```suggestion
    IsShutDown::guard();

    let caller = ic::caller();

    crate::progress().await;

    let ledger = ic::get_mut::<Ledger>();
    let amount_u64: u64 =
        utils::convert_nat_to_u64(amount).expect("transfer failed - unable to convert amount");
    let fee = compute_fee(amount_u64);
    ledger.transfer_from(&caller, &from, &to, amount_u64, fee)?;

    let transaction = Transaction {
        timestamp: ic::time(),
        cycles: amount_u64,
        fee,
        kind: TransactionKind::TransferFrom {
            caller: caller,
            from: from,
            to: to,
        },
    };

    Ok(Nat::from(
        ic::get_mut::<HistoryBuffer>().push(transaction.clone()),
    ))
```